---
title: "Identification of aeSNPs with differential allelic expression (daeSNPs)"
output:
  html_notebook: default
  html_document: default
---

# Introduction

# Setup

## Package Dependencies
```{r}
library(tibble)
library(dplyr)
library(readr)
library(microbenchmark)
library(ggplot2)
library(stringr)
library(latex2exp)
library(purrr)
```

## Paths

```{r cache=TRUE}
# Root path
rootpath <- "/data/jxavier/Science/Analysis/DaeBreastMicroarrays"
# Input folder: contains all data necessary as input for this script.
inputpath <- file.path(rootpath, "/dae_analysis/dae_analysis/output")
# Script folder: R source code
path_scripts <- file.path(rootpath, "/dae_analysis/dae_analysis/R")
# table of genotypes (AA, AB/BA, BB): row is snp, col is sample.
path_genotypes <- file.path (inputpath, "gtypes.91K.non.Imp.csv")
# log2(RNA/DNA) data: row is snp and columns are samples.
path_snp_eratios <- file.path (inputpath, "rnaLR.dnaLR.ratio2.91K.nonImp.csv")
## Output folder: contains data generated by this script.
outpath <- file.path(rootpath, "results")
```

```{r cache=TRUE}
list.files(rootpath)
list.files(path_scripts)
```



## Local code dependencies

#This notebook depends on two R source files: *dae_definition.R* and *plot.R*. The file *dae_definition.R* contains functions to determine how many heterozygous' samples show differential allelic expression. The file *plot.R* contains neat functions to wrap ggplot2 calls to facilitate plotting.

```{r}
source(file.path(path_scripts, "dae_definition.R"))
source(file.path(path_scripts, "plot.R"))
```

###List all functions loaded into the Global Environment:

```{r cache=TRUE}
lsf.str()
```


# Analysis

## Read in input files

###With regards to input data this notebook depends on two data files:

###1. **genotypes**: `r path_genotypes`
###2. **eratios** (expression ratios): `r path_snp_eratios`

```{r cache=TRUE}
# read in the genotypes
  genotypes <- data.table::fread(input = path_genotypes, sep = ",", header = TRUE, showProgress = FALSE)
  data.table::setnames(genotypes, "V1", "snp")
  data.table::setkey(genotypes, snp)
  
  # read in expression ratios
  eratios <- data.table::fread(input = path_snp_eratios, sep = ",", header = TRUE)
  data.table::setnames(eratios, "V1", "snp")
  data.table::setkey(eratios, snp)

```

```{r cache=TRUE}
nsamples <- ncol(genotypes) - 1 # minus one because first column is the 'snp'
nsnps <- length(common_snps(genotypes, eratios))
```

###There are `r nsamples` samples, the number of common SNPs between genotypes and eratios is `r nsnps`. All samples are the same in both the genotypes and eratios objects, and in the same order:

```{r cache=TRUE}
genotypes_samples <- str_match(colnames(genotypes)[-1], "(\\w+)_")[,2]
eratios_samples <- str_match(colnames(eratios)[-1], "(\\w+)_")[,2]
all(genotypes_samples == eratios_samples)

```

## Differential allelic expression criteria for tSNPs

###In this section, we introduce our working definition of Differential Allelic Expression (DAE) for the tSNPs. A tSNP is said to have DAE if a minimum number of heterozygous' samples have DAE. A specific sample shows DAE if the absolute value of the log2 of expression ratios (eratio) exceeds a certain threshold $\alpha$. Hence, a tSNP is said to show DAE if: (1) a minimum absolute number of heterozygous' samples ($\beta$) show DAE, i.e., its expression ratio exceeeds $\alpha$; and (2) a minimum percentage of heterozygous' samples show DAE ($\gamma$).

###So, whether a tSNP shows DAE depends on those three thresholds: DAE threshold ($\alpha$), absolute number of heterozygous' samples with DAE ($\beta$), and percentage of heterozygous' samples with DAE ($\gamma$).

###To understand the impact of the choices of $\alpha$, $\beta$ and $\gamma$, we decided to sweep the {$\alpha$, $\beta$, $\gamma$}-space and count how many tSNPs pass these criteria. We varied these thresholds like so:

1. $\alpha \in \{0, 0.01, 0.02, ..., 2.99, 3\}$
2. $\beta \in \{0, 1, 2, ...,$ `r nsamples` $\}$
3. $\gamma \in \{0, 0.01, ..., 0.99, 1.00\}$

```{r cache=TRUE}
# alpha
dae.thresholds <- seq(from = 0, to = 3, by = 0.01)
# beta
number_het_dae_samples_thresholds <- 0:nsamples
# gamma
proportion_het_dae_het_samples_thresholds <- seq(from = 0, to = 1, by = 0.01)
```

We start by determining how many tSNPs pass the first criterion, i.e., $\alpha$. To do this, we calculate first a data frame for each $\alpha$ value. Each of these data frames contains a summary of the number of heterozygous' samples that show DAE above $\alpha$. Then, we write these data frames to disk.

```{r cache=TRUE}
# this takes a few minutes...
# define file names
filenames <- file.path(outpath, sprintf("table_of_het_samples_showing_dae_%.2f.csv", dae.thresholds))

# define function for writing summaries to files
write_summary <- function(filename, dae_thr, 
                          genotypes, eratios, 
                          comparison = "abs-greater", verbose = TRUE) {
  if(verbose)
    message("Writing ", filename, "...")
  fwrite(summary_of_het_samples_showing_dae(dae_thr, genotypes, eratios, comparison), file = filename)
}
# actually write the summaries to files
invisible(
mapply(write_summary, filenames, dae.thresholds,
       MoreArgs = list(genotypes = genotypes, eratios = eratios, comparison = "abs-greater", verbose = FALSE))
)

```

```{r cache=TRUE}
# read in all summaries into a list
summaries_list <- lapply(filenames, fread)
# assign names to list according to the DAE threshold
names(summaries_list) <- sprintf("%.2f", dae.thresholds)

# generate combinations of the three thresholds
threshold_combinations <- as_tibble(expand.grid( 
            prop_het_dae_thr = proportion_het_dae_het_samples_thresholds,
            number_het_dae_thr = number_het_dae_samples_thresholds,
            dae_thr = names(summaries_list), stringsAsFactors = FALSE)[, 3:1])
head(threshold_combinations)
```


```{r cache=TRUE}
# this takes ~ 30 min (TODO: make it parallel)
number_of_tsnps <-
  mapply(function(dae_thr, number_het_dae_thr, prop_het_dae_thr) {
    number_of_tsnps_with_dae(summaries_list[[dae_thr]], 
                             number_het_dae_thr, prop_het_dae_thr)
  },
  threshold_combinations$dae_thr,
  threshold_combinations$number_het_dae_thr,
  threshold_combinations$prop_het_dae_thr
  )


number_of_tsnps_df <- add_column(threshold_combinations, number_of_tsnps)
fwrite(number_of_tsnps_df, file = file.path(outpath, "tsnp_with_dae.csv"))

```

```{r}
number_of_tsnps_df <- read_csv(file.path(outpath, "tsnp_with_dae.csv"))
```

```{r}
number_of_tsnps_df
```

## Number of tSNPs with DAE

```{r}
dae_reference_values <- round(c(log2(1.2), log2(1.5), log2(2), log2(3)), digits = 2)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1, 5, 10, 15, 20, 30), 0.05, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=2.5, y=80000, label= TeX("$\\gamma = 0.05$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1, 5, 10, 15, 20, 30), 0.15, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=2.5, y=80000, label= TeX("$\\gamma = 0.15$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1, 5, 10, 15, 20, 30), 0.30, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=2.5, y=80000, label= TeX("$\\gamma = 0.30$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1,2,3,4,5,10), 0.05, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=2.5, y=80000, label= TeX("$\\gamma = 0.05$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1,2,3,4,5,10), 0.05, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   xrange = c(0,1),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\gamma = 0.05$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1,2,3,4,5,10), 0.10, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   xrange = c(0,1),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\gamma = 0.10$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "dae_thr", "number_het_dae_thr", "prop_het_dae_thr",
                   c(1,2,3,4,5,10), 0.15, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\alpha,$ DAE threshold"),
                   xrange = c(0,1),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta,$ no. of hets' with DAE")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\gamma = 0.15$", output='character'), parse = TRUE)
```



```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "number_het_dae_thr", "dae_thr", "prop_het_dae_thr",
                   dae_reference_values, 0.05, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\beta,$ no. of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=50, y=80000, label= TeX("$\\gamma = 0.05$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "number_het_dae_thr", "dae_thr", "prop_het_dae_thr",
                   dae_reference_values, 0.15, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\beta,$ no. of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=50, y=80000, label= TeX("$\\gamma = 0.15$", output='character'), parse = TRUE)
```


```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "number_het_dae_thr", "dae_thr", "prop_het_dae_thr",
                   dae_reference_values, 0.30, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\beta,$ no. of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=50, y=80000, label= TeX("$\\gamma = 0.30$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "dae_thr", "number_het_dae_thr",
                   dae_reference_values, 1, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\beta = 1$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "dae_thr", "number_het_dae_thr",
                   dae_reference_values, 2, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\beta = 2$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "dae_thr", "number_het_dae_thr",
                   dae_reference_values, 3, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\beta = 3$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "dae_thr", "number_het_dae_thr",
                   dae_reference_values, 5, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\beta = 5$", output='character'), parse = TRUE)
```

```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "dae_thr", "number_het_dae_thr",
                   dae_reference_values, 10, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\alpha,$ DAE threshold")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\beta = 10$", output='character'), parse = TRUE)
```


```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "number_het_dae_thr", "dae_thr",
                   c(0, 1, 2, 3, 4, 5, 10), 0.59, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta$")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\alpha = 0.58$", output='character'), parse = TRUE) +
  geom_vline(xintercept=0.10)
```


```{r}
snps_with_dae_plot(number_of_tsnps_df, 
                   "prop_het_dae_thr", "number_het_dae_thr", "dae_thr",
                   c(1, 2, 3, 4, 5, 10), 1, 
                   ylabel_left = "No. of tSNPs w/ DAE",
                   ylabel_right = "tSNPs w/ DAE (%)",
                   xlabel = TeX("$\\gamma,$ percentage of hets' with DAE"),
                   yrange = c(NA_real_, nsnps),
                   tsnps_max_100 = nsnps,
                   ind_variable2_label = TeX("$\\beta$")) +
  annotate("text", x=0.8, y=80000, label= TeX("$\\alpha = 1$", output='character'), parse = TRUE)
```
  
"END"

DAE_0.58_dataframe = summaries_list[['0.58']]
dim(DAE_0.58_dataframe)
#[1] 91383     8
write.csv(DAE_0.58_dataframe, file.path(outpath, "DAE_0.58_dataframe"))

DAE_0.58_dataframe_3hets_10percent = subset (DAE_0.58_dataframe, number_het_dae_samples >= 3 & proportion_het_dae_het_samples >=0.10)
dim(DAE_0.58_dataframe_3hets_10percent) 
#[1] 44267     8
write.csv(DAE_0.58_dataframe_3hets_10percent, file.path(outpath, "DAE_0.58_dataframe_3hets_10percent"))

####################


